File: task-redirect.test.mjs
// Behavioral test for configs/opencode/plugin/task-redirect.js — the
// OpenCode-side mirror of configs/claude/task-redirect.sh (see
// task_redirect_test.go, package main, for the bash script's equivalent
// end-to-end test).
//
// This file is deliberately NOT under configs/opencode/plugin/: that
// directory is copied byte-for-byte to the user's OpenCode plugin dir
// (internal/apps/opencode/opencode.go's ForceConfigure step, via
// files.CopyDir) — a test file living there would ship to every installed
// machine. Keeping this test at the repo root (sibling to
// task_redirect_test.go) mirrors the Go test's placement without polluting
// the deployed plugin directory.
//
// Run with: node --test task-redirect.test.mjs
// Uses only Node's built-in test runner and assert module (available since
// Node 18) — no new dependency for a single test file, per CLAUDE.md's
// "prefer the standard library" rule applied to this repo's one JS file.

import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  TaskRedirect,
  isDevgitaRepo,
} from "./configs/opencode/plugin/task-redirect.js";

// This test file lives at the repo root, so its own directory is a devgita
// repo (repo-root go.mod has module github.com/cjairm/devgita). Used as the
// devgita-cwd for the release-gating tests below.
const DEVGITA_DIR = dirname(fileURLToPath(import.meta.url));

// runHook drives the plugin exactly as OpenCode would: build the plugin,
// invoke its tool.execute.before hook with a bash tool call, and report
// whether it denied (threw) or allowed (returned normally). `ctx` is the
// OpenCode plugin context (defaults to the devgita repo dir so the release
// rules are exercised in their firing state unless a test overrides it).
async function runHook(command, env = {}, ctx = { directory: DEVGITA_DIR }) {
  const previous = {};
  for (const [key, value] of Object.entries(env)) {
    previous[key] = process.env[key];
    if (value === undefined) {
      delete process.env[key];
    } else {
      process.env[key] = value;
    }
  }
  try {
    const plugin = await TaskRedirect(ctx);
    const hook = plugin["tool.execute.before"];
    try {
      await hook({ tool: "bash" }, { args: { command } });
      return { denied: false, message: null };
    } catch (err) {
      return { denied: true, message: err.message };
    }
  } finally {
    for (const [key, value] of Object.entries(previous)) {
      if (value === undefined) {
        delete process.env[key];
      } else {
        process.env[key] = value;
      }
    }
  }
}

test("allows legitimate single commands", async () => {
  const allowed = [
    "git diff",
    "git diff HEAD~1",
    "git diff --stat",
    "git log",
    "git log -5",
    "git log --oneline",
    "git tag",
    "git tag v1.0.0",
    "git tag -l",
    "git reset --soft HEAD",
    "git worktree list",
    "git worktree prune",
    "git status",
    "git push origin main",
    'git commit -m "fix: something"',
    // Compound commands where no segment matches any rule.
    "cd some/dir && git status",
    "git fetch && git log -5",
    // A commit message mentioning a trigger word must not itself trigger a
    // rule — "worktree" here is message text, not a `git worktree` call.
    'git commit -m "fix: worktree stuff"',
    // A commit message containing separator-like characters (';', '&&')
    // must not be split apart: the quoted span is one segment, and neither
    // half looks like a git invocation on its own.
    'git commit -m "fix: a; b"',
    // Pathological case for quote-aware splitting: this commit message
    // literally contains "&& git worktree" as text. A naive
    // (non-quote-aware) splitter would slice this into a segment starting
    // with "git worktree" and falsely deny it. Still a single `git commit`
    // command — must be allowed.
    'git commit -m "notes && git worktree stuff"',
    // gh commands that must NOT match the new gh rules: `pr view` is a
    // different subcommand from `pr review`/`pr checks`; `pr status`/`pr list`
    // are neither; a graphql query without reviewThreads and a bare `gh api`
    // are not the review-threads fetch.
    "gh pr view",
    "gh pr view --json title",
    "gh pr status",
    "gh pr list",
    "gh api graphql -f query='{ viewer { login } }'",
    "gh api repos/cjairm/devgita",
  ];
  for (const command of allowed) {
    const result = await runHook(command);
    assert.equal(
      result.denied,
      false,
      `expected allow for ${JSON.stringify(command)}, got deny: ${result.message}`,
    );
  }
});

test("denies narrow patterns, including compound commands and env-var prefixes", async () => {
  const cases = [
    ["git diff main..feature", "devgita task review-package"],
    ["git diff v1.2.0..v1.3.0", "devgita task review-package"],
    ["git diff --stat A..B", "devgita task review-package"],
    ["git log --oneline base..head", "devgita task review-package"],
    ["git worktree add ../wt -b feature-x", "devgita task worktree-start"],
    ["git worktree remove ../wt", "devgita task worktree-finish"],
    // New gh rules — all GLOBAL, so they deny regardless of cwd. (runHook
    // defaults ctx to the devgita dir, but these rules never consult it; the
    // release-gating tests below cover the cwd-dependent rules.)
    ["gh pr checks", "devgita task pr-checks"],
    ["gh pr checks --watch", "devgita task pr-checks"],
    ["gh pr review --approve", "devgita task submit-review"],
    ["gh pr review --request-changes -b bad", "devgita task submit-review"],
    [
      "gh api graphql --paginate -f query='{ repository { pullRequest { reviewThreads { nodes { id } } } } }'",
      "devgita task review-threads",
    ],
    // Compound commands: a matching segment anywhere in the chain must
    // deny, not just a bare command at position 0.
    [
      "cd some/dir && git worktree add ../wt -b x",
      "devgita task worktree-start",
    ],
    ["git status; git worktree remove ../wt", "devgita task worktree-finish"],
    ["git fetch && git diff main..feature", "devgita task review-package"],
    ["gh pr view && gh pr checks", "devgita task pr-checks"],
    // git diff a..b | less: the LHS of the pipe is still a git invocation
    // itself, so this must deny too.
    ["git diff main..feature | less", "devgita task review-package"],
    // Env-var-prefix case (no separator character before `git` at all):
    // deliberately handled now that GIT_PREFIX is being reworked anyway — a
    // simple `NAME=value` prefix, single or repeated, in front of `git`
    // still denies.
    ["GIT_PAGER=cat git diff main..feature", "devgita task review-package"],
    [
      "FOO=bar BAZ=qux git worktree add ../wt -b x",
      "devgita task worktree-start",
    ],
  ];
  for (const [command, wantReplacement] of cases) {
    const result = await runHook(command);
    assert.equal(
      result.denied,
      true,
      `expected deny for ${JSON.stringify(command)}`,
    );
    assert.ok(
      result.message.includes(wantReplacement),
      `expected deny reason for ${JSON.stringify(command)} to mention ${JSON.stringify(wantReplacement)}, got ${JSON.stringify(result.message)}`,
    );
    assert.ok(
      result.message.includes("DEVGITA_SKIP_TASK_REDIRECT"),
      `expected deny reason for ${JSON.stringify(command)} to state the bypass escape hatch, got ${JSON.stringify(result.message)}`,
    );
  }
});

// isDevgitaRepo is unit-tested directly with injected paths so the release
// gate's core decision is verified independently of the plugin plumbing.
test("isDevgitaRepo detects the devgita go.mod by walking up", () => {
  // The repo root itself is devgita.
  assert.equal(isDevgitaRepo(DEVGITA_DIR), true);
  // A nested subdirectory still resolves upward to the devgita go.mod.
  assert.equal(isDevgitaRepo(join(DEVGITA_DIR, "cmd")), true);

  // A dir with no go.mod anywhere up the (temp) tree: not devgita.
  const noGoMod = mkdtempSync(join(tmpdir(), "no-gomod-"));
  assert.equal(isDevgitaRepo(noGoMod), false);

  // A dir whose go.mod is a different module: not devgita.
  const otherModule = mkdtempSync(join(tmpdir(), "other-mod-"));
  writeFileSync(
    join(otherModule, "go.mod"),
    "module github.com/other/thing\n\ngo 1.25\n",
  );
  assert.equal(isDevgitaRepo(otherModule), false);

  // Indeterminate inputs fail toward false (release rules do not fire).
  assert.equal(isDevgitaRepo(undefined), false);
  assert.equal(isDevgitaRepo(""), false);
  assert.equal(isDevgitaRepo(123), false);
});

test("release rules deny inside devgita, allow everywhere else", async () => {
  const releaseCommands = [
    "git reset --soft HEAD~1",
    "git reset --soft HEAD~3",
    "git tag -a v0.12.0 -m release",
    "git tag -a -m release v0.12.0",
    "cd wt && git reset --soft HEAD~2",
    "git status && git tag -a v1.0.0 -m release",
  ];

  const noGoMod = mkdtempSync(join(tmpdir(), "no-gomod-"));
  const otherModule = mkdtempSync(join(tmpdir(), "other-mod-"));
  writeFileSync(
    join(otherModule, "go.mod"),
    "module github.com/other/thing\n\ngo 1.25\n",
  );

  for (const command of releaseCommands) {
    // Inside devgita: deny.
    const inside = await runHook(command, {}, { directory: DEVGITA_DIR });
    assert.equal(
      inside.denied,
      true,
      `expected deny inside devgita for ${JSON.stringify(command)}`,
    );
    assert.ok(inside.message.includes("devgita task release"));

    // Outside devgita (no go.mod, and a different module): allow.
    for (const dir of [noGoMod, otherModule]) {
      const outside = await runHook(command, {}, { directory: dir });
      assert.equal(
        outside.denied,
        false,
        `expected allow outside devgita (${dir}) for ${JSON.stringify(command)}, got: ${outside.message}`,
      );
    }
  }
});

test("release gate uses worktree fallback and fails toward not firing", async () => {
  // No `directory` in ctx: falls back to `worktree`.
  const viaWorktree = await runHook(
    "git reset --soft HEAD~1",
    {},
    { worktree: DEVGITA_DIR },
  );
  assert.equal(viaWorktree.denied, true);

  // Empty ctx and a non-devgita process.cwd() would allow; here we assert the
  // explicit non-devgita directory path allows (the fail-toward-not-firing
  // posture), which is the safety-critical direction.
  const noGoMod = mkdtempSync(join(tmpdir(), "no-gomod-"));
  const failOpen = await runHook(
    "git tag -a v1.2.3 -m r",
    {},
    { directory: noGoMod },
  );
  assert.equal(failOpen.denied, false);
});

test("bypass env var allows everything", async () => {
  const result = await runHook("git worktree add ../wt -b x", {
    DEVGITA_SKIP_TASK_REDIRECT: "1",
  });
  assert.equal(result.denied, false);
});

test("ignores non-bash tool calls and non-string commands", async () => {
  const plugin = await TaskRedirect();
  const hook = plugin["tool.execute.before"];

  // Non-bash tool: never even inspects the command.
  await assert.doesNotReject(() =>
    hook(
      { tool: "edit" },
      { args: { command: "git worktree add ../wt -b x" } },
    ),
  );

  // Missing/non-string command: falls through without throwing.
  await assert.doesNotReject(() => hook({ tool: "bash" }, { args: {} }));
  await assert.doesNotReject(() =>
    hook({ tool: "bash" }, { args: { command: 123 } }),
  );
});

File: task_redirect_test.go
package main

import (
	"bytes"
	"encoding/json"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

// runTaskRedirectHook extracts configs/claude/task-redirect.sh from the
// embedded ConfigsFS (the same bytes that ship in the built binary — see
// TestTmuxDefaultCommandStaysResurrectSafe in embedded_test.go for the same
// against-the-embedded-FS pattern), runs it with a PreToolUse-shaped JSON
// payload on stdin, and returns its exit code and stderr.
//
// The script shells out to jq to parse its stdin payload, exactly as it does
// when Claude Code invokes it — jq is a required runtime dependency of the
// deployed hook (see docs/apps/claude.md), not a mocked external command, so
// running it here tests the shipped script's actual behavior rather than a
// stand-in. If jq isn't on PATH, the test is skipped rather than failing —
// the same posture the hook itself takes if jq is unavailable at runtime
// (documented fail-open behavior), and CI environments that lack it can't
// meaningfully validate this script anyway.
func runTaskRedirectHook(t *testing.T, command string) (exitCode int, stderr string) {
	t.Helper()
	// Default working dir for the process: this repo's own root (a devgita
	// go.mod), reached by running with the test's own cwd. Global rules don't
	// care about it; the release-gating tests below use the cwd-aware helper.
	return runTaskRedirectHookInDir(t, command, "", "")
}

// runTaskRedirectHookInDir runs the hook with an explicit payload `cwd` field
// (payloadCwd, omitted when empty) and an explicit process working directory
// (procDir, inherited when empty). Both feed the release-rule devgita-repo
// gate: the script reads `.cwd` from the payload and falls back to its own
// $PWD, so these two knobs exercise the gate's every input.
func runTaskRedirectHookInDir(
	t *testing.T,
	command, payloadCwd, procDir string,
) (exitCode int, stderr string) {
	t.Helper()

	if _, err := exec.LookPath("jq"); err != nil {
		t.Skip("jq not found on PATH; skipping task-redirect.sh behavioral test")
	}

	scriptBytes, err := fs.ReadFile(ConfigsFS, "configs/claude/task-redirect.sh")
	if err != nil {
		t.Fatalf("failed to read embedded task-redirect.sh: %v", err)
	}

	dir := t.TempDir()
	scriptPath := filepath.Join(dir, "task-redirect.sh")
	if err := os.WriteFile(scriptPath, scriptBytes, 0o755); err != nil {
		t.Fatalf("failed to write script: %v", err)
	}

	payloadMap := map[string]any{
		"tool_input": map[string]string{"command": command},
	}
	if payloadCwd != "" {
		payloadMap["cwd"] = payloadCwd
	}
	payload, err := json.Marshal(payloadMap)
	if err != nil {
		t.Fatalf("failed to marshal payload: %v", err)
	}

	cmd := exec.Command(scriptPath)
	cmd.Stdin = bytes.NewReader(payload)
	if procDir != "" {
		cmd.Dir = procDir
	}
	var stderrBuf bytes.Buffer
	cmd.Stderr = &stderrBuf
	cmd.Stdout = &bytes.Buffer{} // discard; must stay empty per the script's own contract

	runErr := cmd.Run()
	if runErr == nil {
		return 0, stderrBuf.String()
	}
	var exitErr *exec.ExitError
	if !isExitError(runErr, &exitErr) {
		t.Fatalf("failed to run task-redirect.sh for command %q: %v", command, runErr)
	}
	return exitErr.ExitCode(), stderrBuf.String()
}

// repoRoot returns this test binary's repo root (the directory holding the
// devgita go.mod), used as a real devgita-cwd for the release-gating tests.
func repoRoot(t *testing.T) string {
	t.Helper()
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("failed to get working directory: %v", err)
	}
	return wd
}

func isExitError(err error, target **exec.ExitError) bool {
	ee, ok := err.(*exec.ExitError)
	if ok {
		*target = ee
	}
	return ok
}

func TestTaskRedirectHook_AllowsLegitimateSingleCommands(t *testing.T) {
	allowed := []string{
		"git diff",
		"git diff HEAD~1",
		"git diff --stat",
		"git log",
		"git log -5",
		"git log --oneline",
		"git tag",
		"git tag v1.0.0",
		"git tag -l",
		"git reset --soft HEAD",
		"git worktree list",
		"git worktree prune",
		"git status",
		"git push origin main",
		"git commit -m \"fix: something\"",
		// Compound commands where no segment matches any rule.
		"cd some/dir && git status",
		"git fetch && git log -5",
		// A commit message mentioning a trigger word must not itself trigger
		// a rule — the "worktree" here is just message text, not a
		// `git worktree` invocation.
		"git commit -m \"fix: worktree stuff\"",
		// A commit message containing separator-like characters (';', '&&')
		// must not be split apart by the segmenter: the quoted span is one
		// segment, and neither half looks like a git invocation on its own.
		"git commit -m \"fix: a; b\"",
		// Pathological case for quote-aware splitting: this commit message
		// literally contains "&& git worktree" as message text. A naive
		// (non-quote-aware) splitter would slice this into a second segment
		// that starts with "git worktree" and falsely deny it. It's still a
		// single `git commit` command and must be allowed.
		"git commit -m \"notes && git worktree stuff\"",
		// gh commands that must NOT match the new gh rules: `pr view` is a
		// different subcommand from `pr review`/`pr checks`; `pr status`/`pr
		// list` are neither; a graphql query without reviewThreads and a bare
		// `gh api` are not the review-threads fetch.
		"gh pr view",
		"gh pr view --json title",
		"gh pr status",
		"gh pr list",
		"gh api graphql -f query='{ viewer { login } }'",
		"gh api repos/cjairm/devgita",
	}
	for _, command := range allowed {
		t.Run(command, func(t *testing.T) {
			code, stderr := runTaskRedirectHook(t, command)
			if code != 0 {
				t.Errorf(
					"expected allow (exit 0) for %q, got exit %d, stderr=%q",
					command,
					code,
					stderr,
				)
			}
		})
	}
}

func TestTaskRedirectHook_DeniesNarrowPatterns(t *testing.T) {
	cases := []struct {
		command         string
		wantReplacement string
	}{
		{"git diff main..feature", "devgita task review-package"},
		{"git diff v1.2.0..v1.3.0", "devgita task review-package"},
		{"git diff --stat A..B", "devgita task review-package"},
		{"git log --oneline base..head", "devgita task review-package"},
		{"git worktree add ../wt -b feature-x", "devgita task worktree-start"},
		{"git worktree remove ../wt", "devgita task worktree-finish"},
		// New gh rules — all GLOBAL, so they deny regardless of cwd (this
		// helper runs with no payload cwd; the process cwd is the repo root,
		// but these rules never consult it).
		{"gh pr checks", "devgita task pr-checks"},
		{"gh pr checks --watch", "devgita task pr-checks"},
		{"gh pr review --approve", "devgita task submit-review"},
		{"gh pr review --request-changes -b bad", "devgita task submit-review"},
		{
			"gh api graphql --paginate -f query='{ repository { pullRequest { reviewThreads { nodes { id } } } } }'",
			"devgita task review-threads",
		},
		// Compound commands: a matching segment anywhere in the chain must
		// deny, not just a bare command at position 0.
		{"cd some/dir && git worktree add ../wt -b x", "devgita task worktree-start"},
		{"git status; git worktree remove ../wt", "devgita task worktree-finish"},
		{"git fetch && git diff main..feature", "devgita task review-package"},
		{"gh pr view && gh pr checks", "devgita task pr-checks"},
		// git diff a..b | less: the LHS of the pipe is still a git
		// invocation itself, so this must deny too.
		{"git diff main..feature | less", "devgita task review-package"},
		// Env-var-prefix case (no separator character before `git` at all):
		// deliberately handled now that the anchor is being reworked anyway
		// (see GIT_ANCHOR in task-redirect.sh / GIT_PREFIX in
		// task-redirect.js) — a simple `NAME=value` prefix, single or
		// repeated, in front of `git` still denies.
		{"GIT_PAGER=cat git diff main..feature", "devgita task review-package"},
		{"FOO=bar BAZ=qux git worktree add ../wt -b x", "devgita task worktree-start"},
	}
	for _, tc := range cases {
		t.Run(tc.command, func(t *testing.T) {
			code, stderr := runTaskRedirectHook(t, tc.command)
			if code != 2 {
				t.Fatalf(
					"expected deny (exit 2) for %q, got exit %d, stderr=%q",
					tc.command,
					code,
					stderr,
				)
			}
			if !strings.Contains(stderr, tc.wantReplacement) {
				t.Errorf(
					"expected deny reason for %q to mention %q, got %q",
					tc.command,
					tc.wantReplacement,
					stderr,
				)
			}
			if !strings.Contains(stderr, "DEVGITA_SKIP_TASK_REDIRECT") {
				t.Errorf(
					"expected deny reason for %q to state the bypass escape hatch, got %q",
					tc.command,
					stderr,
				)
			}
		})
	}
}

// TestTaskRedirectHook_ReleaseRulesGatedToDevgitaRepo is the regression-proof
// assertion for this fix: the two release rules (git reset --soft HEAD~N, git
// tag -a v<semver>) deny ONLY when the command runs inside the devgita repo,
// and allow the identical command everywhere else. "Inside devgita" means a
// go.mod with module github.com/cjairm/devgita found by walking up from the
// payload's cwd (falling back to the process $PWD). The gate fails toward NOT
// firing, so an indeterminate cwd allows the raw git through.
func TestTaskRedirectHook_ReleaseRulesGatedToDevgitaRepo(t *testing.T) {
	releaseCommands := []string{
		"git reset --soft HEAD~1",
		"git reset --soft HEAD~3",
		"git tag -a v0.12.0 -m release",
		"git tag -a -m release v0.12.0",
		"cd wt && git reset --soft HEAD~2",
		"git status && git tag -a v1.0.0 -m release",
	}

	devgitaDir := repoRoot(t) // this repo's own root has a devgita go.mod

	// A non-devgita dir with no go.mod, and one with a different module path —
	// both must ALLOW the release commands.
	noGoMod := t.TempDir()
	otherModule := t.TempDir()
	if err := os.WriteFile(
		filepath.Join(otherModule, "go.mod"),
		[]byte("module github.com/other/thing\n\ngo 1.25\n"),
		0o644,
	); err != nil {
		t.Fatalf("failed to write other go.mod: %v", err)
	}

	t.Run("devgita cwd denies", func(t *testing.T) {
		for _, command := range releaseCommands {
			t.Run(command, func(t *testing.T) {
				code, stderr := runTaskRedirectHookInDir(t, command, devgitaDir, "")
				if code != 2 {
					t.Fatalf(
						"expected deny (exit 2) inside devgita for %q, got exit %d, stderr=%q",
						command,
						code,
						stderr,
					)
				}
				if !strings.Contains(stderr, "devgita task release") {
					t.Errorf(
						"expected deny reason to mention 'devgita task release', got %q",
						stderr,
					)
				}
			})
		}
	})

	t.Run("non-devgita cwd allows", func(t *testing.T) {
		for _, dir := range []string{noGoMod, otherModule} {
			for _, command := range releaseCommands {
				t.Run(dir+"/"+command, func(t *testing.T) {
					code, stderr := runTaskRedirectHookInDir(t, command, dir, "")
					if code != 0 {
						t.Fatalf(
							"expected allow (exit 0) outside devgita for %q in %q, got exit %d, stderr=%q",
							command,
							dir,
							code,
							stderr,
						)
					}
				})
			}
		}
	})

	t.Run("no cwd field falls back to process PWD and allows outside devgita", func(t *testing.T) {
		// No payload cwd; process cwd is a non-devgita temp dir — the gate's
		// $PWD fallback must resolve to non-devgita and allow.
		for _, command := range releaseCommands {
			t.Run(command, func(t *testing.T) {
				code, stderr := runTaskRedirectHookInDir(t, command, "", noGoMod)
				if code != 0 {
					t.Fatalf(
						"expected allow (exit 0) with no cwd and non-devgita PWD for %q, got exit %d, stderr=%q",
						command,
						code,
						stderr,
					)
				}
			})
		}
	})
}

func TestTaskRedirectHook_FailsOpenOnMalformedInput(t *testing.T) {
	if _, err := exec.LookPath("jq"); err != nil {
		t.Skip("jq not found on PATH")
	}

	scriptBytes, err := fs.ReadFile(ConfigsFS, "configs/claude/task-redirect.sh")
	if err != nil {
		t.Fatalf("failed to read embedded task-redirect.sh: %v", err)
	}
	dir := t.TempDir()
	scriptPath := filepath.Join(dir, "task-redirect.sh")
	if err := os.WriteFile(scriptPath, scriptBytes, 0o755); err != nil {
		t.Fatalf("failed to write script: %v", err)
	}

	stdins := map[string]string{
		"empty stdin":             "",
		"not json":                "not json at all",
		"json without command":    `{"tool_input":{"file_path":"foo.go"}}`,
		"json with empty command": `{"tool_input":{"command":""}}`,
	}
	for name, stdin := range stdins {
		t.Run(name, func(t *testing.T) {
			cmd := exec.Command(scriptPath)
			cmd.Stdin = strings.NewReader(stdin)
			var stderrBuf bytes.Buffer
			cmd.Stderr = &stderrBuf
			if err := cmd.Run(); err != nil {
				t.Errorf(
					"expected the hook to fail open (exit 0) for %s, got error: %v (stderr=%q)",
					name,
					err,
					stderrBuf.String(),
				)
			}
		})
	}
}

func TestTaskRedirectHook_BypassEnvVarAllowsEverything(t *testing.T) {
	if _, err := exec.LookPath("jq"); err != nil {
		t.Skip("jq not found on PATH")
	}

	scriptBytes, err := fs.ReadFile(ConfigsFS, "configs/claude/task-redirect.sh")
	if err != nil {
		t.Fatalf("failed to read embedded task-redirect.sh: %v", err)
	}
	dir := t.TempDir()
	scriptPath := filepath.Join(dir, "task-redirect.sh")
	if err := os.WriteFile(scriptPath, scriptBytes, 0o755); err != nil {
		t.Fatalf("failed to write script: %v", err)
	}

	payload, _ := json.Marshal(map[string]any{
		"tool_input": map[string]string{"command": "git worktree add ../wt -b x"},
	})
	cmd := exec.Command(scriptPath)
	cmd.Stdin = bytes.NewReader(payload)
	cmd.Env = append(os.Environ(), "DEVGITA_SKIP_TASK_REDIRECT=1")
	var stderrBuf bytes.Buffer
	cmd.Stderr = &stderrBuf
	if err := cmd.Run(); err != nil {
		t.Errorf(
			"expected bypass env var to allow a normally-denied command, got error: %v (stderr=%q)",
			err,
			stderrBuf.String(),
		)
	}
}
